File: packages/EarthSciSerialization.jl/src/grid_accessor.jl
# GridAccessor interface (gt-hvl4).
#
# This file defines the *interface* that concrete grid impls must satisfy
# for ESS to query grid geometry during discretization, tree-walk, and
# codegen. Concrete per-family implementations live in the
# EarthSciDiscretizations (ESD) binding per the 2026-04-22 grid-inversion
# decision: ESS owns the trait + registration hook; ESD owns the concrete
# subtypes and registers them on package init.
#
# Signature surface (per EarthSciDiscretizations/docs/GRIDS_API.md):
#   cell_centers(g, idx...)            — cell-center coordinates
#   neighbors(g, cell)                 — adjacent cells
#   metric_eval(g, name, idx...)       — named metric field (dx, dy, area, …)
#
# Index shape is family-dependent:
#   rectilinear 2D:     idx = (i, j)         cell = (i, j)
#   rectilinear 3D:     idx = (i, j, k)      cell = (i, j, k)
#   block_structured:   idx = (panel, i, j)  cell = (panel, i, j)
#   unstructured:       idx = (cell_id,)     cell = cell_id
#
# Options keys are snake_case on the wire (GRIDS_API.md §2.2); this file
# only defines the accessor surface, not the generator surface.

"""
    GridAccessor

Abstract supertype for grid-accessor types. Concrete subtypes live in the
ESD (EarthSciDiscretizations) binding and register themselves via
[`register_grid_accessor!`](@ref).

Concrete `T <: GridAccessor` must implement:

- [`cell_centers`](@ref)`(g::T, idx...)` — cell-center coordinate tuple
- [`neighbors`](@ref)`(g::T, cell)` — iterable of adjacent cells
- [`metric_eval`](@ref)`(g::T, name::AbstractString, idx...)` — metric field value
"""
abstract type GridAccessor end

"""
    GridAccessorError(message)

Thrown by the default (unimplemented) generic stubs and by
[`grid_accessor_factory`](@ref) / [`make_grid_accessor`](@ref) when a
family has no registered factory.
"""
struct GridAccessorError <: Exception
    message::String
end
Base.showerror(io::IO, e::GridAccessorError) = print(io, "GridAccessorError: ", e.message)

"""
    cell_centers(grid::GridAccessor, idx...)

Cell-center coordinates of the cell at `idx`. Return shape is
family-dependent (e.g. `(lon, lat)` for spherical families,
`(x, y, z)` for 3D cartesian).
"""
function cell_centers end

"""
    neighbors(grid::GridAccessor, cell)

Iterable of cells adjacent to `cell`. `cell` uses the index convention
matching the family's `cell_centers` signature.
"""
function neighbors end

"""
    metric_eval(grid::GridAccessor, name::AbstractString, idx...)

Value of the named metric field at `idx`. Valid names are family-defined
(typical: `"dx"`, `"dy"`, `"dz"`, `"area"`, `"volume"`). Implementations
throw [`GridAccessorError`](@ref) for unknown names.
"""
function metric_eval end

# Default unimplemented methods. Subtypes opt in by overriding these with
# more specific signatures; the fallbacks guarantee a clean error with the
# concrete type name instead of a bare `MethodError`.
cell_centers(g::GridAccessor, args...) =
    throw(GridAccessorError("cell_centers not implemented for $(typeof(g))"))
neighbors(g::GridAccessor, cell) =
    throw(GridAccessorError("neighbors not implemented for $(typeof(g))"))
metric_eval(g::GridAccessor, name::AbstractString, args...) =
    throw(GridAccessorError("metric_eval not implemented for $(typeof(g))"))

# ---------------------------------------------------------------------------
# Registration hook: ESD concrete impls call register_grid_accessor! on
# package init. ESS code dispatches into the registry by the RFC §6
# `family` string.
# ---------------------------------------------------------------------------

const _GRID_ACCESSOR_REGISTRY = Dict{String,Any}()

"""
    register_grid_accessor!(family, factory) -> previous_factory_or_nothing

Register `factory` as the constructor for grid family `family`. `factory`
is any callable `factory(grid_data) -> GridAccessor`, where `grid_data`
is the RFC §6 grid dict (e.g. `EsmFile.grids["g1"].data`).

Family names match the RFC §6 enum (`"cartesian"`, `"lat_lon"`,
`"stretched_lat_lon"`, `"cubed_sphere"`, `"stretched_cubed_sphere"`,
`"mpas"`, `"duo"`).

Replacing an existing entry returns the previous factory. This is the
ESS → ESD handshake: ESS owns the trait + registry; ESD registers
concrete subtypes.
"""
function register_grid_accessor!(family::AbstractString, factory)
    key = String(family)
    prev = get(_GRID_ACCESSOR_REGISTRY, key, nothing)
    _GRID_ACCESSOR_REGISTRY[key] = factory
    return prev
end

"""
    unregister_grid_accessor!(family) -> Bool

Remove the factory for `family`. Returns `true` if removed, `false` if
none was registered.
"""
function unregister_grid_accessor!(family::AbstractString)
    return pop!(_GRID_ACCESSOR_REGISTRY, String(family), nothing) !== nothing
end

"""
    grid_accessor_factory(family) -> factory

Return the registered factory for `family`, or throw
[`GridAccessorError`](@ref) if none is registered.
"""
function grid_accessor_factory(family::AbstractString)
    f = get(_GRID_ACCESSOR_REGISTRY, String(family), nothing)
    f === nothing && throw(GridAccessorError(
        "no grid accessor registered for family '$(family)'; " *
        "ensure EarthSciDiscretizations (or an equivalent provider) is loaded"))
    return f
end

"""
    registered_grid_families() -> Vector{String}

Sorted list of grid families with a registered accessor.
"""
registered_grid_families() = sort!(collect(keys(_GRID_ACCESSOR_REGISTRY)))

"""
    make_grid_accessor(family, grid_data) -> GridAccessor

Construct a `GridAccessor` for `family` by dispatching to the registered
factory with `grid_data`.
"""
function make_grid_accessor(family::AbstractString, grid_data)
    return grid_accessor_factory(family)(grid_data)
end

File: packages/earthsci-toolkit/src/grid-accessor.ts
/**
 * Grid accessor contract (ESS-side) for the 2026-04-22 grid-inversion
 * decision. ESS defines the accessor surface and registration hook;
 * concrete family implementations live in EarthSciDiscretizations (ESD)
 * and register themselves with this package at runtime.
 *
 * Method names mirror `EarthSciDiscretizations/docs/GRIDS_API.md` §2 and
 * §3.4: snake_case on the wire and on accessor methods, `camelCase` only
 * for TS-native invariants (`toESM`). See `gt-j2b8`.
 *
 * This module is signatures-only. No family is implemented here — the
 * registry starts empty and is populated by the ESD bindings.
 */

export type Dtype = 'float64' | 'float32'

/**
 * Cell index. Block-structured families use a compound index
 * (`[panel, i, j]`); rectilinear families use `[i, j]` or a flat
 * scalar. The accessor instance defines the convention; consumers
 * pass the index shape the accessor hands out in `neighbors`.
 */
export type CellIndex = number | readonly number[]

/**
 * Geographic / cartesian center of a cell. Spherical families populate
 * `lon`/`lat`; cartesian families populate `x`/`y`/`z`. The accessor
 * may populate both for families that carry both coordinate systems.
 */
export interface CellCenter {
  readonly lon?: number
  readonly lat?: number
  readonly x?: number
  readonly y?: number
  readonly z?: number
}

/**
 * Accessor contract every ESD-provided concrete grid implements.
 * Subsumes the GRIDS_API §3.4 `Grid` interface (family/dtype/toESM)
 * and adds the three accessor methods called out in gt-j2b8.
 */
export interface GridAccessor {
  /** Family name per GRIDS_API §1, e.g. `"cubed_sphere"`. */
  readonly family: string
  /** Element precision. Cross-binding conformance is promised at `"float64"`. */
  readonly dtype: Dtype
  /** Total cell count (interior + ghosts, consistent with `ghosts`). */
  readonly n_cells: number

  /**
   * Cell center at logical index `(i, j)`. For families with more than
   * two logical indices (e.g. cubed_sphere panels), wrap the extra
   * indices in the accessor's factory closure and expose a 2-D view.
   */
  cell_centers(i: number, j: number): CellCenter

  /** Indices of cells adjacent to `cell`, in family-defined order. */
  neighbors(cell: CellIndex): readonly CellIndex[]

  /**
   * Evaluate a named metric field (e.g. `"area"`, `"jac"`, `"dx"`) at
   * logical index `(i, j)`. Field names are family-specific; an
   * unknown name is a family-level error.
   */
  metric_eval(name: string, i: number, j: number): number

  /**
   * Lower to a §6-schema-valid ESM object. After `canonicalize()`
   * (§5.4.6), the result is the cross-binding conformance ground
   * truth (GRIDS_API §3.5).
   */
  toESM(): object
}

/**
 * Factory function a concrete family registers. Given the options
 * object (the same snake_case keys the `.esm` wire form carries),
 * construct and return a ready-to-use accessor.
 */
export type GridAccessorFactory = (opts: Record<string, unknown>) => GridAccessor

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export const E_GRID_FAMILY_UNKNOWN = 'E_GRID_FAMILY_UNKNOWN'
export const E_GRID_FAMILY_ALREADY_REGISTERED = 'E_GRID_FAMILY_ALREADY_REGISTERED'
export const E_GRID_FAMILY_NAME_INVALID = 'E_GRID_FAMILY_NAME_INVALID'

export class GridAccessorError extends Error {
  readonly code: string
  constructor(code: string, message?: string) {
    super(message ?? code)
    this.code = code
    this.name = 'GridAccessorError'
  }
}

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

const registry = new Map<string, GridAccessorFactory>()

function assertValidFamilyName(name: string): void {
  if (typeof name !== 'string' || name.length === 0) {
    throw new GridAccessorError(
      E_GRID_FAMILY_NAME_INVALID,
      'grid family name must be a non-empty string',
    )
  }
}

/**
 * Register a factory for `family`. Re-registering the same family is
 * an error so downstream code cannot silently pick the wrong
 * implementation when two packages ship the same name. Use
 * `unregisterGridFamily` first if you need to swap.
 */
export function registerGridFamily(family: string, factory: GridAccessorFactory): void {
  assertValidFamilyName(family)
  if (typeof factory !== 'function') {
    throw new GridAccessorError(
      E_GRID_FAMILY_NAME_INVALID,
      `factory for grid family "${family}" must be a function`,
    )
  }
  if (registry.has(family)) {
    throw new GridAccessorError(
      E_GRID_FAMILY_ALREADY_REGISTERED,
      `grid family "${family}" is already registered`,
    )
  }
  registry.set(family, factory)
}

/**
 * Remove a registration. Returns `true` iff the family was present.
 * Intended for tests and ESD hot-reload; not a production path.
 */
export function unregisterGridFamily(family: string): boolean {
  assertValidFamilyName(family)
  return registry.delete(family)
}

/** Look up a previously-registered factory. */
export function getGridFamily(family: string): GridAccessorFactory | undefined {
  return registry.get(family)
}

/** `true` iff a factory is registered under `family`. */
export function hasGridFamily(family: string): boolean {
  return registry.has(family)
}

/**
 * All registered family names, sorted lexicographically so output is
 * deterministic across runs.
 */
export function listGridFamilies(): string[] {
  return Array.from(registry.keys()).sort()
}

/**
 * Construct a grid via the registry. Raises
 * `GridAccessorError(E_GRID_FAMILY_UNKNOWN)` if the family has no
 * registered factory.
 */
export function createGrid(family: string, opts: Record<string, unknown> = {}): GridAccessor {
  const factory = registry.get(family)
  if (factory === undefined) {
    throw new GridAccessorError(
      E_GRID_FAMILY_UNKNOWN,
      `grid family "${family}" is not registered; call registerGridFamily() first`,
    )
  }
  return factory(opts)
}

/**
 * Drop every registered family. Intended for test isolation; do not
 * call in production code.
 */
export function clearGridFamilies(): void {
  registry.clear()
}
